rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRACDIExtension.java

```diff
@@ -0,0 +1,249 @@
+
+/**
+ * This CDI extension collects all LRA participants that contain
+ * one or more non-JAX-RS participant methods. The collected classes are stored
+ * in {@link LRAParticipantRegistry}.
+ */
+public class LRACDIExtension implements Extension {
+
+    private ClassPathIndexer classPathIndexer = new ClassPathIndexer();
+    private final Map<String, LRAParticipant> participants = new HashMap<>();
+
+    public void observe(@Observes AfterBeanDiscovery event, BeanManager beanManager) throws IOException, ClassNotFoundException {
+        Index index = classPathIndexer.createIndex();
+
+        List<AnnotationInstance> annotations = index.getAnnotations(DotName.createSimple("javax.ws.rs.Path"));
+
+        for (AnnotationInstance annotation : annotations) {
+            ClassInfo classInfo;
+            AnnotationTarget target = annotation.target();
+
+            if (target.kind().equals(AnnotationTarget.Kind.CLASS)) {
+                classInfo = target.asClass();
+            } else if (target.kind().equals(AnnotationTarget.Kind.METHOD)) {
+                classInfo = target.asMethod().declaringClass();
+            } else {
+                continue;
+            }
+
+            Class<?> clazz = getClass().getClassLoader().loadClass(classInfo.name().toString());
+
+            LRAParticipant participant = processJavaClass(clazz);
+            if (participant != null) {
+                participants.put(participant.getId().getName(), participant);
+                Set<Bean<?>> participantBeans = beanManager.getBeans(clazz, new AnnotationLiteral<Any>() {});
+                if (participantBeans.isEmpty()) {
+                    // resource is not registered as managed bean so register a custom managed instance
+                    try {
+                        participant.setInstance(clazz.newInstance());
+                    } catch (InstantiationException | IllegalAccessException e) {
+                        LRALogger.i18NLogger.error_cannotProcessParticipant(e);
+                    }
+                }
+            }
+        }
+
+        event.addBean()
+            .read(beanManager.createAnnotatedType(LRAParticipantRegistry.class))
+            .beanClass(LRAParticipantRegistry.class)
+            .scope(ApplicationScoped.class)
+            .createWith(context -> new LRAParticipantRegistry(participants));
+    }
+
+    /**
+     * Collects all non-JAX-RS participant methods in the defined Java class
+     *
+     * @param javaClass a class to be scanned
+     * @return Collected methods wrapped in {@link LRAParticipant} class or null if no non-JAX-RS methods have been found
+     */
+    private LRAParticipant processJavaClass(Class<?> javaClass) {
+        if (isNotLRAParticipant(javaClass)) {
+            return null;
+        }
+
+        LRAParticipant participant = new LRAParticipant(javaClass);
+
+        Arrays.stream(javaClass.getDeclaredMethods()).forEach(m -> processParticipantMethod(m, participant));
```

Comment:
Didn't you just stream all the methods in the call to isNotLRAParticipant - maybe you could do both jobs in one pass.

Reply:
Actually this code is ran once at deployment time so performance is less off an issue. Feel free to ignore this comment if you don't want to refactor the code to improve its performance.

Reply:
Thanks, I agree. It will be ran only once during deployment and the benefit is that we won't process (verify signatures and save references) for classes that are not participants (the first pass through declared methods).